qiskit/algorithms/eigen_solvers/numpy_eigen_solver.py

```diff
@@ -123,7 +124,18 @@ def _solve(self, operator: OperatorBase) -> None:
             for i, idx in enumerate(indices):
                 eigvec[idx, i] = 1.0
         else:
-            if self._k >= 2 ** operator.num_qubits - 1:
+            if (
+                isinstance(operator, PauliSumOp)
+                and operator.is_hermitian()
+                and self._k >= 2 ** operator.num_qubits - 1
+            ):
+                logger.debug("SciPy doesn't support to get all eigenvalues, using NumPy instead.")
+                eigval, eigvec = np.linalg.eigh(operator.to_matrix())
+            elif isinstance(operator, PauliSumOp) and operator.is_hermitian():
+                eigval, eigvec = scisparse.linalg.eigsh(
+                    operator.to_spmatrix(), k=self._k, which="SR"
+                )
```

Comment:
Would it be more effiicient, to have 
```
if (instance(operator, PauliSumOp) and operator.is_hermitian()):
    if (self._k >=2.....
        ...
    else:
        ...
elif self._k >=2 ...
   ...
else:
  ...
```
So the is_hermitian() test can only be done once.

I assume it would also be possible, instead of checking specifically for PauliSumOp, which is known to have the is_hermitian() method - would be just to check the operator to see if it has that method. This way if it were ever added for other operators its a bit more future proof.

Reply:
Thank you. It can certainly improve efficiency. I'll fix it.

I think it is nice to bind it with Protocol for the future.

Reply:
After much thought, I finally decided to add the method to OperatorBase. If we have an operator that has performance problems, we should probably define efficient methods in that class.

Reply:
Ok, hopefully we do not go via that too often - most commonly used is the PauliSumOp. Otherwise the base class does a matrix conv for the hermitian test and then its converted to a matrix again for the eigenvalue extraction.